`std/strutils`: `split`, `rsplit` now raise on an empty separator

ZoomRmc opened the pull request:
This PR adds a `ValueError` raise on an empty separators for `split` and `rsplit` iterators and functions. 
This fixes an infinite loop for an empty separator.

Previous code was inconsistent as only one function of all variations had a forced assertion for this case.

Raising an exception was chosen over an assertion following the behaviour of Python's [`str.split`](https://docs.python.org/3/library/stdtypes.html#str.split).

Current behaviour:
```nim
import std/[strutils, enumerate]

# returns the input string
doAssert "setsAreSpecial".split({}) == @["setsAreSpecial"]

# infinite loop
for i, _ in enumerate("nineChars".split("")):
  if i > 8:
    echo "boom"
    break
  else:
    stdout.write i+1, ","
# 1,2,3,4,5,6,7,8,9,boom
```

ZoomRmc committed:
`strutils.split` now raises on an empty separator

juancarlospaco commented:
Maybe try just checking the `set[char]` arguments only, to not break packages.

ZoomRmc commented:
Well, it doesn't make any sense to have different behaviour depending on the type of the separator collection used.

The alternatives:
1. Replace checks with forced asserts, this is not breaking but shifts the burden of checking to the caller. Not great, but fixes the infinite loop bug.

2. Change implementation to return input string on empty separator. 
  It could be argued that an empty separator is not exactly a logic error. One example is a dynamic value coming from a user implying "don't split at all".

3. Change implementation to work like Rust's [`string::split`](https://doc.rust-lang.org/std/primitive.str.html#method.split):
    > When the empty string is used as a separator, it separates every character in the string, along with the beginning and end of the string.
      ```nim
      doAssert "nim".split("") == @["", "n", "i", "m", ""]
      ```

juancarlospaco commented:
4)
Make a new one with a `{.since.}` that works correctly, in-place, with `out` parameters, kinda like 
`func split*(s: string; sep: char; result: out seq[string]) = `
so you do not break any packages because is a new symbol, old one can be Deprecated later
(user can always outplace with `sugar.dup` anyways).

arnetheduck commented:
Raising a new exception is a breaking change indeed, to be avoided - the rust version seems to achieve the same thing without breaking working code?

ZoomRmc commented:
Regarding [juancarlospaco](https://github.com/juancarlospaco)'s suggestion, I think introducing a weird replacement with the correct behaviour, to outlive the broken but expected version is just asking for another round of deprecation later on. Sure way to keep maintainers of tomorrow busy.

Regarding Rust version: It doesn't break anything just because it introduces a valid output for a case which has always resulted in a broken code previously, and (we hope) no one relied on generating broken code with this. However, is this output for this kind of input in any way useful? Is there any algorithm relying on generating empty strings for start and end of a split string? To me it looks just like a codified by-product of the specific implementation Rust used. If anyone knows otherwise, let me know. 

The most logical output is returning the input string, as it's what you get when you "split by no element".

I'm not a fan of exceptions, byt `ValueError` at least provides a way of handling this specific scenario (not that anyone cares for it, it seems). It's also consistent with the general approach of the stdlib and follows Python's [`split`](https://hg.python.org/cpython/file/tip/Objects/stringlib/split.h#l154), which many people see as a valuable way to handle stuff.

Regarding breaking changes: well, isn't it a devel branch? There's plenty of time to fix code and add another exception to track. If the exception is not listed it doesn't mean the code works, so even though packages built fine, they still had this bug lurking. Not many people track exceptions (sadly), and at least those who don't will be getting an exception and not an infinite loop with this PR.

juancarlospaco commented:
What do you mean " weird " ?, `out` parameters are not weird nor deprecated,
you can reuse the same `result` without allocating a new seq. 😐

ZoomRmc commented:
> What do you mean " weird " ?, `out` parameters are not weird nor deprecated, you can reuse the same `result` without allocating a new seq. :neutral_face:

It's weird in this specific context. I'm sorry, but I don't really understand what are the possible scenarios where reusing a seq for splits with an out parameter is beneficial but that can't be properly handled by the compiler with a regular assignment syntax.  

Out parameters are great for side-effects, foreign code compatibility or when you simply want to be explicit that reusing a buffer is the way to go. In this case it would be just a hacky solution, in my opinion. It very well may be I'm not seeing the full picture. Besides, while not deprecated, they are still experimental.

juancarlospaco commented:
The point was to create a new overload for `split` with other arguments so it wont break packages.

ZoomRmc commented:
It's not like this change breaks the world. Mostly a few packages where people were careful enough to track exceptions. It's not hard to fix. Besides, no one has suggested to backport this change to stable versions.

Is it really that critical to afford keeping two versions (broken and not broken) in the stdlib?

I'd like to hear thoughts on other alternatives as well.

arnetheduck commented:
> Mostly a few packages

this is churn that needlessly takes these packages off the market until someone invests yet more time into fixing them instead of focusing on new things - this is represents a cost to the community that slows down development for anyone that wants to use nim seriously - it is significant for anyone that wants to write significant amounts of code.

The other point here is that it's _annoying_ to have an exception raised from simple utilities like this which _can_ be given a reasonable behavior - ie "split every letter" like rust does on empty input is quite sensible - one way to treat text is that between each character there is an implicit "0-length" marker that "logic" can latch on to, and that is what it does.

The only thing an annoying exception on this edge case leads to is that it annoys those that want to use exception tracking because now they have to try/except around every rsplit.

Given how long this has been in the std lib, it's also not the case that it happens very often that it's called this way - clearly, it doesn't matter that much which behavior is chosen as long as it can reasonably be defended (whether that be return each character which is "useful" or return the whole string (in which case you can simply not call split)..

Araq commented:
Only the following options are acceptable:

- Document the behaviour as "undefined".
- Add an `assert` which is thankfully not tracked by the exception tracking.
- Do define the result to be some concrete value. This is what Rust does.

I prefer the first option as everything else has costs: Asserts are usually not disabled so everybody pays for the code size increase and Rust's solution encourages the JavaScript programmer not to learn how to do string handling as `split` is the only thing a JS devs knows and misuses for everything everywhere.

ZoomRmc commented:
I don't like doing it the Rust way (a rare case for me) because `strutils.split` is not unicode-aware and we shouldn't pretend it is. Splitting by a "0-length marker" encourages using split instead of iterating or indexing a string, giving a hint that it's somehow different when it's not. So I agree with @Araq, it will be misused.

I think returning a `@["inputString"]`, documenting this behaviour and mentioning iterating/indexing as a solution for splitting between the chars would be preferable. Slapping asserts is fine too.

"Undefined" is a punishment for not reading the docs. I get the slight sadistic appeal of it, but the problem is that undefined behaviour punishes the users much more than developers. No bueno.

arnetheduck commented:
> I think returning a @["inputString"], documenting this behaviour and mentioning iterating/indexing as a solution for splitting between the chars would be preferable.

SGTM - any behavior for this edge case is better than a dumb exception nobody will want to catch :)

Araq commented:
> "Undefined" is a punishment for not reading the docs. I get the slight sadistic appeal of it, but the problem is that undefined behaviour punishes the users much more than developers. No bueno.

That's one way of looking at it, but here is another way: It always was UB and if the documentation finally says so, this knowledge carries over to other, older Nim versions in wide usage. We didn't introduce the UB deliberately in order to setup a trap, it just happened to be the case that it always was UB.

arnetheduck commented:
>  It always was UB and if the documentation finally says so,

Have the documentation say that it used to be UB, if this is a concern
